### PR TITLE
Use C++14 in CMake and related fixes

### PR DESCRIPTION
**Pull request by cyyever**
I just change CMAKE_CXX_STANDARD to 14.

```
build_image:Custom=centos:7
buildworker:Custom=linux-6
```

### CODE REVIEW

File: cmake/OpenCVDetectCXXCompiler.cmake [hunk 1]
```diff
@@ -194,54 +194,7 @@ if(CMAKE_VERSION VERSION_LESS "3.1")
   endforeach()
 endif()
 
-if(NOT OPENCV_SKIP_CMAKE_CXX_STANDARD)
-  ocv_update(CMAKE_CXX_STANDARD 11)
```
Comment (alalek) [inline cmake/OpenCVDetectCXXCompiler.cmake:198]: Nope.

Existed code allows compilation even with C++20.
Your patch breaks that and forces C++14.

File: cmake/OpenCVDetectCXXCompiler.cmake [hunk 2]
```diff
@@ -194,54 +194,7 @@ if(CMAKE_VERSION VERSION_LESS "3.1")
...
-
-if((HAVE_CXX11 AND OPENCV_ENABLE_ATOMIC_LONG_LONG
-        AND NOT MSVC
-        AND NOT (X86 OR X86_64)
-    AND NOT OPENCV_SKIP_LIBATOMIC_COMPILER_CHECK)
-    OR OPENCV_FORCE_LIBATOMIC_COMPILER_CHECK
-)
-  ocv_check_compiler_flag(CXX "" HAVE_CXX_ATOMICS_WITHOUT_LIB "${OpenCV_SOURCE_DIR}/cmake/checks/atomic_check.cpp")
-  if(NOT HAVE_CXX_ATOMICS_WITHOUT_LIB)
-    list(APPEND CMAKE_REQUIRED_LIBRARIES atomic)
```
Comment (alalek) [inline cmake/OpenCVDetectCXXCompiler.cmake:234]: Until you have reproducer for corresponding issue please do not change checks. Not all configuration cases are tested in these 10 public builders.

Use `git blame` (GitHub was Web UI for that) to find corresponding commits, its notes, PRs and linked issues.

**Comment by cyyever** [comment]
@alalek The previous change breaked BC. I try to keep the check now.

**Comment by asenyaev** [comment]
jenkins cn please retry a build

**Comment by asenyaev** [comment]
jenkins cn please retry a build

**Comment by HattrickGenerator** [comment]
Does this mean we can use C++14 is becoming the minimal required standard for compiling opencv 5.x?
I.e. we can rely on C++14 for implementing functionality in opencv 5.x?

**Comment by cyyever** [comment]
@HattrickGenerator  Although this is my own proposal, it is reasonable to set CXX14 as default.

**Comment by HattrickGenerator** [comment]
> @HattrickGenerator Although this is my own proposal, it is reasonable to set CXX14 as default.

@cyyever I'm still waiting for the official answer to this, as https://github.com/opencv/opencv/pull/21450 my pull request is much easier to implement with C++14.

**Pull request closed without merging**